docs/design_proposals/transparent-init.md

```diff
@@ -0,0 +1,94 @@
+# Transparent Init
+
+* Author(s): Marlon Gamez
+* Design Shepherd: \<skaffold-core-team-member\>
+* Date: 10/13/2020
+* Status: [Reviewed/Cancelled/Under implementation/Complete]
+
+## Background
+
+This proposal is brought about by an older issue, [#1273](https://github.com/GoogleContainerTools/skaffold/issues/1273)
+
+With the recent focus on skaffold UX, we've decided to see how we can improve the onboarding experience when using skaffold.
+One way we'd like to do this is by allowing the user to simply install skaffold and run any of the skaffold commands in their project, without having to first run `skaffold init` and go through the interactive portion of creating a config for their project.
+To do this, we'd like to automatically create a config upon invocation of skaffold commands, so that the user doesn't have to see something like this
+```
+❯ skaffold dev
+skaffold config file skaffold.yaml not found - check your current working directory, or try running `skaffold init`
+```
+when trying to run skaffold.
+
+## Design
+
+With transparent init, a user's first run of `skaffold dev` could look something like this:
+
+Given a simple project structure:
+```
+project-dir
+├── Dockerfile
+├── README.md
+├── k8s-pod.yaml
+└── main.go
+```
+Running skaffold dev:
+```
+❯ skaffold dev
+Skaffold config file not found. Generating temporary config for use this run.
+Listing files to watch...
+ - skaffold-example
+Generating tags...
+ - skaffold-example -> skaffold-example:v1.14.0-59-g7e5ae4cbc-dirty
+Checking cache...
+ - skaffold-example: Not found. Building
+Found [minikube] context, using local docker daemon.
+Building [skaffold-example]...
+Sending build context to Docker daemon  3.072kB
+...
+```
+
+There are two approaches to implementation that I've considered:
+
+**Creating a temporary `skaffold.yaml` by running `skaffold init --force`**
+
+Under the hood view:
+- `skaffold dev` is run
+- skaffold searches for a `skaffold.yaml` file in the directory, finds nothing
+- skaffold creates a new config using `skaffold init --force`
+- the new config is parsed and brought into memory
+- the config file is deleted from disk
+- `skaffold dev` continues like normal
+
+**Creating the config in memory using init functions**
+
+Under the hood view:
+- `skaffold dev` is run
+- skaffold searches for a `skaffold.yaml` file in the directory, finds nothing
+- skaffold creates a new config and keeps in memory
+- `skaffold dev` continues like normal
+
+## Open Issues/Questions
+
+**Should generation of the config all happen in memory? By refactoring `DoInit()` we could probably generate the config and start using it without having to write/read the `skaffold.yaml` from disk**
+
+*Unresolved*
+
+**Should we prompt the user to ask if they'd like to save this config for future use?**
+
+*Unresolved*
```

Comment:
Not completely sure how I feel about this, because we'd be writing a file out without the user's consent/knowledge beforehand. When a user runs `skaffold init` by itself it would make sense that they expect a file to appear on disk, but I can't say the same about `skaffold dev`

Reply:
After giving this some more thought, I think we should just write it as long as the message we give to the user makes it clear that we're doing this. We should make them aware that they can just remove it if unwanted or run `skaffold init` if they want to have some more control over what they get.